README.md
# Pager

Numeric pagination suitable for multiple cases, such as, custom paged table.

<p align="center">
  <img src="https://media.giphy.com/media/r8zXyoxgJVgEQpGYLe/giphy.gif">
  
</p>
<p align="center">
  <img src= "https://media3.giphy.com/media/jCE3XUgsvzhyq31tQJ/giphy.gif">
</p>

## Getting started

You should ensure that you add the pager as a dependency in your flutter project.
```yaml
dependencies:
 pager: ^0.0.6
```

## Example Project

There is a example project in the `example` folder. Check it out. Otherwise, keep reading to get up and running.

## Usage

Need to include the import the package to the dart file where it will be used, use the below command,

```dart
import 'package:pager/pager.dart';

```
**Pager**

Basic Widget

```dart

new Pager(
            currentPage: _currentPage,
            totalPages: 5,
            onPageChanged: (page) {},
          ),

```

Example 1

```dart
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Pagination Example"),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Container(
              height: 200,
              margin: const EdgeInsets.all(32),
              color: Colors.blue,
              child: Center(
                child: Text(
                  "PAGE: $_currentPage",
                  style: const TextStyle(color: Colors.white, fontSize: 20),
                ),
              ),
            ),
            Pager(
              currentPage: _currentPage,
              totalPages: 5,
              onPageChanged: (page) {
                setState(() {
                  _currentPage = page;
                });
              },
            ),
          ],
        ),
      ),
    );
  }
}
```

Complete example 2

```dart
import 'package:flutter/material.dart';
import 'package:pager/pager.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Pagination Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late int _currentPage;
  late int _itemsPerPage;
  List<int> pages = [5, 10, 18, 50, 100];
  late int _totalPages;
  final int _totalElements = 100;
  @override
  void initState() {
    super.initState();
    _currentPage = 1;
    _itemsPerPage = pages.first;
    _totalPages = (_totalElements / _itemsPerPage) as int;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Pagination Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Pagination Example'),
        ),
        body: Padding(
          padding: const EdgeInsets.all(32.0),
          child: Center(
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                Container(
                  height: 180,
                  width: 400,
                  color: Colors.blue,
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.spaceAround,
                    children: [
                      Text(
                        "TOTAL ELEMENT: $_totalElements",
                        style:
                            const TextStyle(color: Colors.white, fontSize: 18),
                      ),
                      Text(
                        "PAGE: $_currentPage",
                        style:
                            const TextStyle(color: Colors.white, fontSize: 18),
                      ),
                      Text(
                        "ITEMS PER PAGE: $_itemsPerPage",
                        style:
                            const TextStyle(color: Colors.white, fontSize: 18),
                      ),
                      Text(
                        "TOTAL PAGES: $_totalPages",
                        style:
                            const TextStyle(color: Colors.white, fontSize: 18),
                      ),
                    ],
                  ),
                ),
                const SizedBox(height: 32),
                Pager(
                  currentItemsPerPage: _itemsPerPage,
                  currentPage: _currentPage,
                  totalPages: _totalPages,
                  onPageChanged: (page) {
                    setState(() {
                      _currentPage = page;
                    });
                  },
                  showItemsPerPage: true,
                  onItemsPerPageChanged: (itemsPerPage) {
                    setState(() {
                      _itemsPerPage = itemsPerPage;
                      _totalPages = (_totalElements / _itemsPerPage).ceil();
                      _currentPage = 1;
                    });
                  },
                  itemsPerPageList: pages,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## 0.0.1

* Initial release.

## 0.0.2

* Readme updated. Example added. Package renamed. Customizations added.

## 0.0.3

* Fix pagesView behavior.

## 0.0.4

* ItemsPerPage Component added.

## 0.0.5

* Current ItemsPerPage fixed.

## 0.0.6

* Fix: Arrow buttons disable.
* Dart SDK constraint changed to ">=2.17.0 <4.0.0"

lib/src/pager.dart
import 'package:flutter/material.dart';
import 'package:pager/src/items_per_page.dart';

// ignore: must_be_immutable
class Pager extends StatefulWidget {
  Pager({
    Key? key,
    required this.totalPages,
    required this.onPageChanged,
    this.showItemsPerPage = false,
    this.currentItemsPerPage,
    this.onItemsPerPageChanged,
    this.itemsPerPageList,
    this.pagesView = 3,
    this.currentPage = 1,
    this.numberButtonSelectedColor = Colors.blue,
    this.numberTextSelectedColor = Colors.white,
    this.numberTextUnselectedColor = Colors.black,
    this.pageChangeIconColor = Colors.grey,
    this.itemsPerPageText,
    this.itemsPerPageTextStyle,
    this.dropDownMenuItemTextStyle,
  })  : assert(currentPage > 0 && totalPages > 0 && pagesView > 0,
            "Fatal Error: Make sure the currentPage, totalPages and pagesView fields are greater than zero. "),
        super(key: key) {
    if (showItemsPerPage) {
      assert(
          currentItemsPerPage != null &&
              onItemsPerPageChanged != null &&
              itemsPerPageList != null &&
              itemsPerPageList!.isNotEmpty,
          "Fatal error: OnItemsPerPageChanged must be implemented or itemsPerPageList is null or empty.");
    }
  }

  /// How many page numbers selectable to show at once.
  int pagesView;

  /// Total pages.
  final int totalPages;

  /// The callback that is called when the page is changed.
  final Function(int) onPageChanged;

  /// Show items per page
  bool showItemsPerPage;

  /// Current items per page.
  final int? currentItemsPerPage;

  /// Items per page list. Example: [5,10,20,50]
  List<int>? itemsPerPageList;

  /// The callback that is called when the page is changed.
  final Function(int)? onItemsPerPageChanged;

  /// Current page. Default is 1.
  int currentPage;

  // Button color of the selected page number.
  final Color numberButtonSelectedColor;

  // Text color of the selected page number.
  final Color numberTextUnselectedColor;

  // Text color of the unselected page numbers.
  final Color numberTextSelectedColor;

  // Color of the next, previous, first and last page icons.
  final Color pageChangeIconColor;

  // ItemsPerPage label text.
  final String? itemsPerPageText;

  // ItemsPerPage label text style.
  final TextStyle? itemsPerPageTextStyle;

  // DropDownButtonMenuItem text style.
  final TextStyle? dropDownMenuItemTextStyle;

  @override
  State<Pager> createState() => _PagerState();
}

class _PagerState extends State<Pager> {
  @override
  Widget build(BuildContext context) {
    pagesViewValidation();
    return Column(
      children: [
        Row(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            IconButton(
              tooltip: "First Page",
              onPressed: widget.currentPage > 1
                  ? () {
                      setState(() {
                        widget.currentPage = 1;
                        widget.onPageChanged(widget.currentPage);
                      });
                    }
                  : null,
              splashRadius: 25,
              icon: Icon(
                Icons.first_page,
                color: widget.pageChangeIconColor,
              ),
            ),
            IconButton(
              tooltip: "Previous",
              onPressed: widget.currentPage > 1
                  ? () {
                      setState(() {
                        widget.currentPage =
                            widget.currentPage > 1 ? widget.currentPage - 1 : 1;
                        widget.onPageChanged(widget.currentPage);
                      });
                    }
                  : null,
              splashRadius: 25,
              icon: Icon(
                Icons.chevron_left,
                color: widget.pageChangeIconColor,
              ),
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                for (int i = getPageStart(getPageEnd()); i < getPageEnd(); i++)
                  TextButton(
                    onPressed: () {
                      setState(() {
                        widget.currentPage = i;
                        widget.onPageChanged(widget.currentPage);
                      });
                    },
                    style: ElevatedButton.styleFrom(
                        shape: const CircleBorder(),
                        backgroundColor: widget.currentPage == i
                            ? widget.numberButtonSelectedColor
                            : null),
                    child: Text(
                      "$i",
                      style: TextStyle(
                        color: widget.currentPage == i
                            ? widget.numberTextSelectedColor
                            : widget.numberTextUnselectedColor,
                      ),
                    ),
                  ),
              ],
            ),
            IconButton(
              tooltip: "Next Page",
              onPressed: widget.currentPage < widget.totalPages
                  ? () {
                      setState(() {
                        widget.currentPage =
                            widget.currentPage < widget.totalPages
                                ? widget.currentPage + 1
                                : widget.totalPages;
                        widget.onPageChanged(widget.currentPage);
                      });
                    }
                  : null,
              splashRadius: 25,
              icon: Icon(
                Icons.chevron_right,
                color: widget.pageChangeIconColor,
              ),
            ),
            IconButton(
              tooltip: "Last Page",
              onPressed: widget.currentPage < widget.totalPages
                  ? () {
                      setState(() {
                        widget.currentPage = widget.totalPages;
                        widget.onPageChanged(widget.currentPage);
                      });
                    }
                  : null,
              splashRadius: 25,
              icon: Icon(
                Icons.last_page,
                color: widget.pageChangeIconColor,
              ),
            ),
          ],
        ),
        if (widget.showItemsPerPage)
          ItemsPerPage(
            currentItemsPerPage: widget.currentItemsPerPage!,
            itemsPerPage: widget.itemsPerPageList!,
            onChanged: widget.onItemsPerPageChanged!,
            itemsPerPageText: widget.itemsPerPageText,
            itemsPerPageTextStyle: widget.itemsPerPageTextStyle,
            dropDownMenuItemTextStyle: widget.dropDownMenuItemTextStyle,
          ),
      ],
    );
  }

  /// Get last page to show in pagination.
  int getPageEnd() {
    return widget.currentPage + widget.pagesView > widget.totalPages
        ? widget.totalPages + 1
        : widget.currentPage + widget.pagesView;
  }

  /// Get first page to show in pagination.
  int getPageStart(int pageEnd) {
    return pageEnd == widget.totalPages + 1
        ? pageEnd - widget.pagesView
        : widget.currentPage;
  }

  /// Validation of pagesView field
  void pagesViewValidation() {
    if (widget.totalPages < widget.pagesView) {
      widget.pagesView = widget.totalPages;
    }
  }
}


lib/pagination.dart
library pagination;

import 'package:flutter/material.dart';

// ignore: must_be_immutable
class Pagination extends StatefulWidget {
  Pagination({
    Key? key,
    required this.totalPages,
    required this.onPageChanged,
    this.pagesView = 3,
    this.currentPage = 1,
    this.numberButtonSelectedColor = Colors.blue,
    this.numberTextSelectedColor = Colors.white,
    this.numberTextUnselectedColor = Colors.black,
    this.pageChangeIconColor = Colors.grey,
  })  : assert(currentPage > 0 && totalPages > 0 && pagesView > 0,
            "Fatal Error: Make sure the currentPage, totalPages and pagesView fields are greater than zero. "),
        super(key: key);

  /// How many page numbers selectable to show at once.
  final int pagesView;

  /// Total pages.
  final int totalPages;

  /// The callback that is called when the page is changed.
  final Function(int) onPageChanged;

  /// Current page. Default is 1.
  int currentPage;

  // Button color of the selected page number.
  final Color numberButtonSelectedColor;

  // Text color of the selected page number.
  final Color numberTextUnselectedColor;

  // Text color of the unselected page numbers.
  final Color numberTextSelectedColor;

  // Color of the next, previous, first and last page icons.
  final Color pageChangeIconColor;

  @override
  State<Pagination> createState() => _PaginationState();
}

class _PaginationState extends State<Pagination> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: [
        IconButton(
          tooltip: "First Page",
          onPressed: () {
            setState(() {
              widget.currentPage = 1;
              widget.onPageChanged(widget.currentPage);
            });
          },
          splashRadius: 25,
          icon: Icon(
            Icons.first_page,
            color: widget.pageChangeIconColor,
          ),
        ),
        IconButton(
          tooltip: "Previous",
          onPressed: () {
            setState(() {
              widget.currentPage =
                  widget.currentPage > 1 ? widget.currentPage - 1 : 1;
              widget.onPageChanged(widget.currentPage);
            });
          },
          splashRadius: 25,
          icon: Icon(
            Icons.chevron_left,
            color: widget.pageChangeIconColor,
          ),
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            for (int i = getPageStart(getPageEnd()); i < getPageEnd(); i++)
              TextButton(
                onPressed: () {
                  setState(() {
                    widget.currentPage = i;
                    widget.onPageChanged(widget.currentPage);
                  });
                },
                style: ElevatedButton.styleFrom(
                    shape: const CircleBorder(),
                    primary: widget.currentPage == i
                        ? widget.numberButtonSelectedColor
                        : null),
                child: Text(
                  "$i",
                  style: TextStyle(
                    color: widget.currentPage == i
                        ? widget.numberTextSelectedColor
                        : widget.numberTextUnselectedColor,
                  ),
                ),
              ),
          ],
        ),
        IconButton(
          tooltip: "Next Page",
          onPressed: () {
            setState(() {
              widget.currentPage = widget.currentPage < widget.totalPages
                  ? widget.currentPage + 1
                  : widget.totalPages;
              widget.onPageChanged(widget.currentPage);
            });
          },
          splashRadius: 25,
          icon: Icon(
            Icons.chevron_right,
            color: widget.pageChangeIconColor,
          ),
        ),
        IconButton(
          tooltip: "Last Page",
          onPressed: () {
            setState(() {
              widget.currentPage = widget.totalPages;
              widget.onPageChanged(widget.currentPage);
            });
          },
          splashRadius: 25,
          icon: Icon(
            Icons.last_page,
            color: widget.pageChangeIconColor,
          ),
        ),
      ],
    );
  }

  /// Get last page to show in pagination.
  int getPageEnd() {
    return widget.currentPage + widget.pagesView > widget.totalPages
        ? widget.totalPages + 1
        : widget.currentPage + widget.pagesView;
  }

  /// Get first page to show in pagination.
  int getPageStart(int pageEnd) {
    return pageEnd == widget.totalPages + 1
        ? pageEnd - widget.pagesView
        : widget.currentPage;
  }
}


lib/src/items_per_page.dart
import 'package:flutter/material.dart';

class ItemsPerPage extends StatefulWidget {
  const ItemsPerPage({
    Key? key,
    required this.currentItemsPerPage,
    required this.itemsPerPage,
    required this.onChanged,
    this.itemsPerPageText,
    this.itemsPerPageTextStyle,
    this.dropDownMenuItemTextStyle,
  }) : super(key: key);
  final int currentItemsPerPage;
  final List<int> itemsPerPage;
  final Function(int) onChanged;
  final String? itemsPerPageText;
  final TextStyle? itemsPerPageTextStyle, dropDownMenuItemTextStyle;

  @override
  State<ItemsPerPage> createState() => _ItemsPerPageState();
}

class _ItemsPerPageState extends State<ItemsPerPage> {
  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        Text(
          widget.itemsPerPageText ?? "Items per page: ",
          style: widget.itemsPerPageTextStyle ??
              const TextStyle(
                color: Colors.grey,
              ),
        ),
        const SizedBox(width: 16),
        DropdownButton(
          value: widget.currentItemsPerPage,
          focusColor: Colors.transparent,
          items: widget.itemsPerPage.map((value) {
            return DropdownMenuItem<int>(
              value: value,
              child: Text(
                value.toString(),
                style: widget.dropDownMenuItemTextStyle ??
                    const TextStyle(
                      fontSize: 14,
                    ),
              ),
            );
          }).toList(),
          onChanged: (value) {
            setState(() {
              widget.onChanged(value as int);
            });
          },
        ),
      ],
    );
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:pager/pager.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Pagination Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late int _currentPage;

  @override
  void initState() {
    super.initState();
    _currentPage = 1;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Pagination Example"),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Container(
              height: 200,
              margin: const EdgeInsets.all(32),
              color: Colors.blue,
              child: Center(
                child: Text(
                  "PAGE: $_currentPage",
                  style: const TextStyle(color: Colors.white, fontSize: 20),
                ),
              ),
            ),
            Pager(
              currentPage: _currentPage,
              totalPages: 5,
              onPageChanged: (page) {
                setState(() {
                  _currentPage = page;
                });
              },
            ),
          ],
        ),
      ),
    );
  }
}
